openmdao/components/exec_comp.py

```diff
@@ -222,28 +205,110 @@ def __init__(self, exprs=[], **kwargs):
+        """
+        Register a callable to be used within an ExecComp expression.
+
+        Parameters
+        ----------
+        name : str
+            Name of the callable.
+        callable_obj : callable
+            The callable.
+        complex_safe : bool
+            If True, the given callable works correctly with complex numbers.
+        """
+        global _expr_dict, _not_complex_safe
+
+        if not callable(callable_obj):
+            raise TypeError(f"{cls.__name__}: '{name}' passed to register() of type "
+                            f"'{type(callable_obj).__name__}' is not callable.")
+        if name in _expr_dict:
+            raise NameError(f"{cls.__name__}: '{name}' has already been registered.")
+
+        if name in _disallowed_names:
+            raise NameError(f"{cls.__name__}: cannot register name '{name}' because "
+                            "it's a reserved keyword.")
+
+        if '.' in name:
+            raise NameError(f"{cls.__name__}: cannot register name '{name}' because "
+                            "it contains '.'.")
+        _expr_dict[name] = callable_obj
+
+        if not complex_safe:
+            _not_complex_safe.add(name)
+
     def setup(self):
         """
         Set up variable name and metadata lists.
         """
+        global _not_complex_safe
+
         if not self._exprs:
             raise RuntimeError("%s: No valid expressions provided to ExecComp(): %s."
                                % (self.msginfo, self._exprs))
-        outs = set()
-        allvars = set()
         exprs = self._exprs
         kwargs = self._kwargs
 
         units = self.options['units']
         shape = self.options['shape']
+        shape_by_conn = self.options['shape_by_conn']
+
+        if shape is not None and shape_by_conn:
+            raise RuntimeError(f"{self.msginfo}: Can't set both shape and shape_by_conn.")
+
+        outs = set()
+        allvars = set()
+
+        exprs_info = [(self._parse_for_out_vars(expr.split('=', 1)[0]),
+                       self._parse_for_names(expr)) for expr in exprs]
+
+        self._requires_fd = {}
 
         # find all of the variables and which ones are outputs
-        for expr in exprs:
-            lhs, _ = expr.split('=', 1)
-            outs.update(self._parse_for_out_vars(lhs))
-            allvars.update(self._parse_for_vars(expr))
+        for i, (onames, names) in enumerate(exprs_info):
+            outs.update(onames)
+            allvars.update(names[0])
+            if _not_complex_safe.intersection(names[1]):
+                for o in onames:
+                    self._requires_fd[o] = names
+
+        if self._requires_fd:
+            inps = []
+            for out, (rhsvars, funcs) in self._requires_fd.items():
+                iset = rhsvars.difference(outs)
+                self._requires_fd[out] = (iset, funcs)
+                inps.extend(iset)
+            self._no_check_partials = False
```

Comment:
I wonder if `_no_check_partials` should be set to False even if  complex_safe was set to True for everything that was registered, in case they are mistaken about it. Might be the only way they can find out if it's some external thing that squelches the casting warnings.